File: Chapter I/1.7/example_pg12.cpp
#include <iostream>
#include <random>
#include <vector>

class DiscreteUniform {
    public: explicit DiscreteUniform(std::vector<int> values)
    : values_(std::move(values)),
    rng_(std::random_device{}()),
    dist_(0, static_cast<int>(values_.size()) - 1) {}

    int draw() {
        return values_[dist_(rng_)];
    }

private:
    std::vector<int> values_;
    std::mt19937 rng_;
    std::uniform_int_distribution<int> dist_;
};

int main() {
    std::vector<int> dice_values;
    for (int i = 1; i <= 6; ++i) dice_values.push_back(i);

    DiscreteUniform P_dice(dice_values);
    std::cout << "P_dice = Uniform(1..6)\n";
    for (int i=0; i<2; ++i) {
        std::cout << (i+1) << "th throw: " << P_dice.draw() << "\n";
    }
    return 0;
}

File: Chapter I/1.7/example_pg12.go
package main

import (
	"fmt"
	"math/rand"
	"time"
)

type DiscreteUniform struct {
	values[] int
}

func NewDiscreteUniform(values []int) *DiscreteUniform{
	v := make([]int, len(values))
	copy(v, values)
	return &DiscreteUniform{values: v}
}

func ( d *DiscreteUniform) Draw() int {
	return d.values[rand.Intn(len(d.values))]
}

func main() {
	rand.Seed(time.Now().UnixNano())
	diceValues := make([]int, 0, 6)
	for i := 1; i <= 6; i++{
		diceValues = append(diceValues, i)
	}

	Pdice := NewDiscreteUniform(diceValues)

	fmt.Println("P_dice = Uniform(1..6)")

	for i := 0; i < 2; i++ {
		fmt.Printf("%dth throw: %d\n", i+1, Pdice.Draw())
	}
}